# Data Cleaning

### Documentation
1. Dataset source: https://www.kaggle.com/datasets/wordsforthewise/lending-club/data

2. Target variable defined as a binary outcome 0/1 based on known credit outcome. 
 - Fully Paid as 0, and 
 - Charged Off, Default, and Late (31-120 days) as 1

2. Missing values

3. Data types

4. Leakage columns

5. Duplicates

6. Outliers

## Import libraries and Load dataset

In [10]:
# Import libraries
import os
import pandas as pd 
import numpy as np
import time

In [30]:
# Define directory path
BASE_DIR = os.path.dirname(os.path.abspath(os.getcwd()))

# Define path to output folder
OUTPUT_DIR = os.path.join(BASE_DIR, "outputs")

# Print
print(f"Base directory: {BASE_DIR}")

Base directory: c:\Users\Nafis\BuildersLab_projects\Teams\Credit-Risk-Default


In [58]:
# Display all columns in the output 
pd.set_option('display.max_columns', None) 
pd.set_option('display.max_colwidth', None)
pd.set_option('display.width', None)

In [6]:
"""import sys
!{sys.executable} -m pip install kagglehub"""


'import sys\n!{sys.executable} -m pip install kagglehub'

In [7]:
# Import the dataset from Kaggle
import kagglehub

# Download latest version
path = kagglehub.dataset_download("wordsforthewise/lending-club")

# Check the path where the dataset files are stored
print("Path to dataset files:", path)

# Check the files that are available in the dataset
print(os.listdir(path))

Path to dataset files: C:\Users\Nafis\.cache\kagglehub\datasets\wordsforthewise\lending-club\versions\3
['accepted_2007_to_2018q4.csv', 'accepted_2007_to_2018Q4.csv.gz', 'rejected_2007_to_2018q4.csv', 'rejected_2007_to_2018Q4.csv.gz']


Note: Kaggle returns 4 different files:
- accepted_2007_to_2018q4.csv
- accepted_2007_to_2018Q4.csv.gz
- rejected_2007_to_2018q4.csv
- rejected_2007_to_2018Q4.csv.gz

The .gz is the compressed version, and the .csv is the same data uncompressed. We're only considering "accepted" data becasue it contains all features/variables needed for the analysis and modeling.

In [ ]:
# Start timer to measure the time taken to load the dataset
start_time = time.time()

# Load dataset
df = pd.read_csv(os.path.join(path, "accepted_2007_to_2018Q4.csv.gz"), low_memory=False)

# End timer and print the time taken to load the dataset
end_time = time.time()
print(f"Time taken to load the dataset: {end_time - start_time:.2f} seconds")

## Data Cleaning
### 1. Data Inspection
### 2. Duplicate Values
### 3. Missing Values 
#### 3.1. Missing Values Summary
#### 3.2. Missing Values Handling

### 1. Data Inspection
a. Shape
b. Data types
c. Data types and missing count
d. Summary statistics

In [9]:
# Check the shape of the dataset
print("Shape of the dataset:", df.shape)

Shape of the dataset: (2260701, 151)


Note: The dataset has 2260701 rows and 151 columns.

In [11]:
# Check a sampel of the dataset
print("Sample of the dataset:")
df.sample(5)

Sample of the dataset:


,id,member_id,loan_amnt,funded_amnt,funded_amnt_inv,term,int_rate,installment,grade,sub_grade,...,hardship_payoff_balance_amount,hardship_last_payment_amount,disbursement_method,debt_settlement_flag,debt_settlement_flag_date,settlement_status,settlement_date,settlement_amount,settlement_percentage,settlement_term
40385,65864893,NaN,8400.0,8400.0,8400.0,36 months,10.64,273.58,B,B4,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
1660319,103427925,NaN,6700.0,6700.0,6700.0,36 months,11.49,220.91,B,B5,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
2125582,121863739,NaN,8000.0,8000.0,8000.0,36 months,10.91,261.57,B,B4,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
531523,120056325,NaN,30000.0,30000.0,30000.0,60 months,18.06,762.79,D,D2,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN
542953,119380457,NaN,5000.0,5000.0,5000.0,36 months,7.21,154.87,A,A3,...,NaN,NaN,Cash,N,NaN,NaN,NaN,NaN,NaN,NaN


In [13]:
# Display all columns
print("Columns in the dataset:")
print(df.columns.tolist())

Columns in the dataset:
['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title', 'emp_length', 'home_ownership', 'annual_inc', 'verification_status', 'issue_d', 'loan_status', 'pymnt_plan', 'url', 'desc', 'purpose', 'title', 'zip_code', 'addr_state', 'dti', 'delinq_2yrs', 'earliest_cr_line', 'fico_range_low', 'fico_range_high', 'inq_last_6mths', 'mths_since_last_delinq', 'mths_since_last_record', 'open_acc', 'pub_rec', 'revol_bal', 'revol_util', 'total_acc', 'initial_list_status', 'out_prncp', 'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int', 'total_rec_late_fee', 'recoveries', 'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d', 'last_fico_range_high', 'last_fico_range_low', 'collections_12_mths_ex_med', 'mths_since_last_major_derog', 'policy_code', 'application_type', 'annual_inc_joint', 'dti_joint', 'verification_status_j

In [18]:
# Check info
print("Info about the dataset:")
df.info(verbose=True, show_counts=True) # Display all columns and count of non-null values

Info about the dataset:
<class 'pandas.DataFrame'>
RangeIndex: 2260701 entries, 0 to 2260700
Data columns (total 151 columns):
 #    Column                                      Non-Null Count    Dtype  
---   ------                                      --------------    -----  
 0    id                                          2260701 non-null  str    
 1    member_id                                   0 non-null        float64
 2    loan_amnt                                   2260668 non-null  float64
 3    funded_amnt                                 2260668 non-null  float64
 4    funded_amnt_inv                             2260668 non-null  float64
 5    term                                        2260668 non-null  str    
 6    int_rate                                    2260668 non-null  float64
 7    installment                                 2260668 non-null  float64
 8    grade                                       2260668 non-null  str    
 9    sub_grade                      

In [25]:
# Display summary statistics for numeric columns
df.describe()

,member_id,loan_amnt,funded_amnt,funded_amnt_inv,int_rate,installment,annual_inc,dti,delinq_2yrs,fico_range_low,fico_range_high,inq_last_6mths,mths_since_last_delinq,mths_since_last_record,open_acc,pub_rec,revol_bal,revol_util,total_acc,out_prncp,out_prncp_inv,total_pymnt,total_pymnt_inv,total_rec_prncp,total_rec_int,total_rec_late_fee,recoveries,collection_recovery_fee,last_pymnt_amnt,last_fico_range_high,last_fico_range_low,collections_12_mths_ex_med,mths_since_last_major_derog,policy_code,annual_inc_joint,dti_joint,acc_now_delinq,tot_coll_amt,tot_cur_bal,open_acc_6m,open_act_il,open_il_12m,open_il_24m,mths_since_rcnt_il,total_bal_il,il_util,open_rv_12m,open_rv_24m,max_bal_bc,all_util,total_rev_hi_lim,inq_fi,total_cu_tl,inq_last_12m,acc_open_past_24mths,avg_cur_bal,bc_open_to_buy,bc_util,chargeoff_within_12_mths,delinq_amnt,mo_sin_old_il_acct,mo_sin_old_rev_tl_op,mo_sin_rcnt_rev_tl_op,mo_sin_rcnt_tl,mort_acc,mths_since_recent_bc,mths_since_recent_bc_dlq,mths_since_recent_inq,mths_since_recent_revol_delinq,num_accts_ever_120_pd,num_actv_bc_tl,num_actv_rev_tl,num_bc_sats,num_bc_tl,num_il_tl,num_op_rev_tl,num_rev_accts,num_rev_tl_bal_gt_0,num_sats,num_tl_120dpd_2m,num_tl_30dpd,num_tl_90g_dpd_24m,num_tl_op_past_12m,pct_tl_nvr_dlq,percent_bc_gt_75,pub_rec_bankruptcies,tax_liens,tot_hi_cred_lim,total_bal_ex_mort,total_bc_limit,total_il_high_credit_limit,revol_bal_joint,sec_app_fico_range_low,sec_app_fico_range_high,sec_app_inq_last_6mths,sec_app_mort_acc,sec_app_open_acc,sec_app_revol_util,sec_app_open_act_il,sec_app_num_rev_accts,sec_app_chargeoff_within_12_mths,sec_app_collections_12_mths_ex_med,sec_app_mths_since_last_major_derog,deferral_term,hardship_amount,hardship_length,hardship_dpd,orig_projected_additional_accrued_interest,hardship_payoff_balance_amount,hardship_last_payment_amount,settlement_amount,settlement_percentage,settlement_term
count,0.0,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260664e+06,2.258957e+06,2.260639e+06,2.260668e+06,2.260668e+06,2.260638e+06,1.102166e+06,359156.000000,2.260639e+06,2.260639e+06,2.260668e+06,2.258866e+06,2.260639e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260668e+06,2.260523e+06,580775.000000,2260668.0,1.207100e+05,120706.000000,2.260639e+06,2.190392e+06,2.190392e+06,1.394538e+06,1.394539e+06,1.394539e+06,1.394539e+06,1.350744e+06,1.394539e+06,1.191818e+06,1.394539e+06,1.394539e+06,1.394539e+06,1.394320e+06,2.190392e+06,1.394539e+06,1.394538e+06,1.394538e+06,2.210638e+06,2.190322e+06,2.185733e+06,2.184597e+06,2.260523e+06,2.260639e+06,2.121597e+06,2.190391e+06,2.190391e+06,2.190392e+06,2.210638e+06,2.187256e+06,519701.000000,1.965233e+06,740359.000000,2.190392e+06,2.190392e+06,2.190392e+06,2.202078e+06,2.190392e+06,2.190392e+06,2.190392e+06,2.190391e+06,2.190392e+06,2.202078e+06,2.107011e+06,2.190392e+06,2.190392e+06,2.190392e+06,2.190237e+06,2.185289e+06,2.259303e+06,2.260563e+06,2.190392e+06,2.210638e+06,2.210638e+06,2.190392e+06,1.080200e+05,108021.000000,108021.000000,108021.000000,108021.000000,108021.000000,106184.000000,108021.000000,108021.000000,108021.000000,108021.000000,35942.000000,10917.0,10917.000000,10917.0,10917.000000,8651.000000,10917.000000,10917.000000,34246.000000,34246.000000,34246.000000
mean,NaN,1.504693e+04,1.504166e+04,1.502344e+04,1.309283e+01,4.458068e+02,7.799243e+04,1.882420e+01,3.068792e-01,6.985882e+02,7.025884e+02,5.768354e-01,3.454092e+01,72.312842,1.161240e+01,1.975278e-01,1.665846e+04,5.033770e+01,2.416255e+01,4.206891e+03,4.205965e+03,1.208256e+04,1.206439e+04,9.505772e+03,2.431388e+03,1.518453e+00,1.438791e+02,2.398257e+01,3.429346e+03,6.876610e+02,6.755397e+02,1.814580e-02,44.164220,1.0,1.236246e+05,19.251817,4.147942e-03,2.327317e+02,1.424922e+05,9.344199e-01,2.779407e+00,6.764314e-01,1.562752e+00,2.122236e+01,3.550665e+04,6.914098e+01,1.290133e+00,2.749923e+00,5.806393e+03,5.703230e+01,3.457394e+04,1.012867e+00,1.47

### 2. Duplicate Values

In [27]:
# Check for duplicate rows
print("Number of duplicate rows:", df.duplicated().sum())

# Check for duplicate values in the "id" column
print("Number of duplicate values in 'id' column:", df['id'].duplicated().sum())

Number of duplicate rows: 0
Number of duplicate values in 'id' column: 0


Note: The dataset has no duplicates.

### 3. Missing Values
#### 3.1. Missing Values Summary

In [33]:
# Summary of missing values
missing_values = df.isnull().sum()
missing_values_percentage = (missing_values / len(df)) * 100

# Create a DataFrame to display the summary of missing values
missing_values_summary = pd.DataFrame({
    'Missing Values': missing_values,
    'Percentage (%)': missing_values_percentage
}).sort_values(by='Missing Values', ascending=False)

# Add 'column_name' as a column in the DataFrame
missing_values_summary['Column names'] = missing_values_summary.index

# Save to a csv file to outputs folder
missing_values_summary.to_csv(os.path.join(OUTPUT_DIR, "missing_values_summary.csv"))

print("Summary of missing values:")
missing_values_summary

Summary of missing values:


,Missing Values,Percentage (%),Column names
member_id,2260701,100.000000,member_id
orig_projected_additional_accrued_interest,2252050,99.617331,orig_projected_additional_accrued_interest
hardship_reason,2249784,99.517097,hardship_reason
hardship_payoff_balance_amount,2249784,99.517097,hardship_payoff_balance_amount
hardship_last_payment_amount,2249784,99.517097,hardship_last_payment_amount
...,...,...,...
total_rec_int,33,0.001460,total_rec_int
disbursement_method,33,0.001460,disbursement_method
hardship_flag,33,0.001460,hardship_flag
debt_settlement_flag,33,0.001460,debt_settlement_flag


In [52]:
# Merge data_dictionary with missing values summary

# Load the data dictionary
data_dictionary = pd.read_csv(os.path.join(BASE_DIR, "data/processed/dictionary_accepted.csv"))

# Merge the data dictionary with the missing values summary
data_dictionary_summary = pd.merge(data_dictionary, missing_values_summary, left_on='Variable', right_on='Column names', how='left')

# Drop the redundant column
data_dictionary_summary = data_dictionary_summary.drop(columns=['Column names'])

# Create a new column for different missingness labels: "Empty (100%)", "Very High (95-99%)", "High (50-94%)", "Moderate (10-49%)", "Low (<10%)"
data_dictionary_summary['Missingness Label'] = pd.cut(
    data_dictionary_summary['Percentage (%)'],
    bins=[-1, 0, 10, 50, 95, 100],
    labels=['None (0%)', 
            'Low (<10%)', 
            'Moderate (10-49%)', 
            'High (50-94%)', 
            'Very High (95-100%)']
)

# Sort the data dictionary summary by Percentage of missing values in descending order
data_dictionary_summary = data_dictionary_summary.sort_values(by='Percentage (%)', ascending=False)

# Save to a csv file to outputs folder
data_dictionary_summary.to_csv(os.path.join(OUTPUT_DIR, "data_dictionary_summary.csv"), index=False)
print("Data dictionary with missing values summary:")
data_dictionary_summary

Data dictionary with missing values summary:


,Variable,Description,Missing Values,Percentage (%),Missingness Label
44,member_id,A unique LC assigned Id for the borrower member.,2260701,100.000000,Very High (95-100%)
140,orig_projected_additional_accrued_interest,The original projected additional interest amount that will accrue for the given hardship payment plan as of the Hardship Start Date. This field will be null if the borrower has broken their hardship payment plan.,2252050,99.617331,Very High (95-100%)
134,hardship_start_date,The start date of the hardship plan period,2249784,99.517097,Very High (95-100%)
141,hardship_payoff_balance_amount,The payoff balance amount as of the hardship plan start date,2249784,99.517097,Very High (95-100%)
142,hardship_last_payment_amount,The last payment amount as of the hardship plan start date,2249784,99.517097,Very High (95-100%)
...,...,...,...,...,...
92,sub_grade,LC assigned loan subgrade,33,0.001460,Low (<10%)
143,disbursement_method,"The method by which the borrower receives their loan. Possible values are: CASH, DIRECT_PAY",33,0.001460,Low (<10%)
128,hardship_flag,Flags whether or not the borrower is on a hardship plan,33,0.001460,Low (<10%)
144,debt_settlement_flag,"Flags whether or not the borrower, who has charged-off, is working with a debt-settlement company.",33,0.001460,Low (<10%)


#### 3.2. Missing Values Handling

1. Column `member_id` is 100% missing, so it can be dropped immediately
2. Data with 95-100% missing

In [54]:
# Display very high missingness columns
very_high_missingness = data_dictionary_summary[data_dictionary_summary['Missingness Label'] == 'Very High (95-100%)']

print("Count of columns with very high missingness (95-100%):", len(very_high_missingness))
print("Columns with very high missingness (95-100%):")
very_high_missingness

Count of columns with very high missingness (95-100%): 34
Columns with very high missingness (95-100%):


,Variable,Description,Missing Values,Percentage (%),Missingness Label
44,member_id,A unique LC assigned Id for the borrower member.,2260701,100.000000,Very High (95-100%)
140,orig_projected_additional_accrued_interest,The original projected additional interest amount that will accrue for the given hardship payment plan as of the Hardship Start Date. This field will be null if the borrower has broken their hardship payment plan.,2252050,99.617331,Very High (95-100%)
134,hardship_start_date,The start date of the hardship plan period,2249784,99.517097,Very High (95-100%)
141,hardship_payoff_balance_amount,The payoff balance amount as of the hardship plan start date,2249784,99.517097,Very High (95-100%)
142,hardship_last_payment_amount,The last payment amount as of the hardship plan start date,2249784,99.517097,Very High (95-100%)
130,hardship_reason,Describes the reason the hardship plan was offered,2249784,99.517097,Very High (95-100%)
129,hardship_type,Describes the hardship plan offering,2249784,99.517097,Very High (95-100%)
131,hardship_status,"Describes if the hardship plan is active, pending, canceled, completed, or broken",2249784,99.517097,Very High (95-100%)
133,hardship_amount,The interest payment that the borrower has committed to make each month while they are on a hardship plan,2249784,99.517097,Very High (95-100%)
132,deferral_term,Amount of months that the borrower is expected to pay less than the contractual monthly payment amount due to a hardship plan,2249784,99.517097,Very High (95-100%)


Note:
- `member_id` is a unique Lending Club assigned ID for the borrower. It was removed from the public dataset for privacy reasons and contains no data.

- Joint application fields such as `revol_bal_joint`, and all `sec_app_*` columns are only available for joint applications (when two people apply for a loan together). Less than ~5% of loans in this dataset are joint applications, so these fields are empty for the remaining 95%.

- Hardship fields such as `hardship_type`, `hardship_reason`, `hardship_status`, `hardship_amount`, `hardship_start_date`, `hardship_end_date`, `hardship_length`, `hardship_dpd`, `hardship_loan_status`, `hardship_payoff_balance_amount`, `hardship_last_payment_amount`, `orig_projected_additional_accrued_interest`, `payment_plan_start_date`, and `deferral_term` describe Lending Club's hardship relief program, which allows struggling borrowers to temporarily pay less than their contractual monthly amount. Only ~0.5% of borrowers enrolled in this program, so these fields are empty for the remaining 99.5%. Note that hardship is for borrowers who are struggling but have not yet defaulted. 

- Settlement fields such as `settlement_status`, `settlement_amount`, `settlement_term`, `settlement_percentage`, `settlement_date`, and `debt_settlement_flag_date` describe cases where Lending Club agreed to accept less than the full amount owed, typically on charged-off loans. Only ~1.5% of loans entered a settlement plan, so these fields are empty for the remaining 98.5%. Note that settlement fields are specifically for loans that have already been charged off.

In [ ]:
# Display high missingness columns
high_missingness = data_dictionary_summary[data_dictionary_summary['Missingness Label'] == 'High (50-94%)']

print("Count of columns with high missingness (50-94%):", len(high_missingness))
print("Columns with high missingness (50-94%):")
high_missingness

Count of columns with high missingness (50-94%): 10
Columns with high missingness (50-94%):


,Variable,Description,Missing Values,Percentage (%),Missingness Label
113,verification_status_joint,"Indicates if the co-borrowers' joint income was verified by LC, not verified, or if the income source was verified",2144971,94.880791,High (50-94%)
17,dti_joint,"A ratio calculated using the co-borrowers' total monthly payments on the total debt obligations, excluding mortgages and the requested LC loan, divided by the co-borrowers' combined self-reported monthly income",2139995,94.660683,High (50-94%)
5,annual_inc_joint,The combined self-reported annual income provided by the co-borrowers during registration,2139991,94.660506,High (50-94%)
15,desc,Loan description provided by the borrower,2134636,94.423632,High (50-94%)
52,mths_since_last_record,The number of months since the last public record.,1901545,84.113069,High (50-94%)
55,mths_since_recent_bc_dlq,Months since most recent bankcard delinquency,1741000,77.011511,High (50-94%)
51,mths_since_last_major_derog,Months since most recent 90-day or worse rating,1679926,74.309960,High (50-94%)
57,mths_since_recent_revol_delinq,Months since most recent revolving delinquency.,1520342,67.250910,High (50-94%)
58,next_pymnt_d,Next scheduled payment date,1345343,59.509993,High (50-94%)
50,mths_since_last_delinq,The number of months since the borrower's last delinquency.,1158535,51.246715,High (50-94%)


Note:
- Joint application fields such as `verification_status_joint`, `dti_joint`, and `annual_inc_joint` are only available for joint applications (when two people apply for a loan together). Less than ~5% of loans in this dataset are joint applications, so these fields are empty for the remaining 95%.

- `desc` is the loan description provided by the borrower. It is free text and optional, and most borrowers did not fill it in. Even where it exists, it is unstructured text that is not usable without NLP, which is out of scope for this project.

- `next_pymnt_d` is the next scheduled payment date. It only exists for active loans, and since we filtered to resolved loans only, this field is mostly empty. It is also a post-origination field and a leakage risk. Note that once a note is charged off, there are no more scheduled patments. Lending Club stops expecting paymenys at that point. For fully paid loans, there is no next playment either since loan is closed.

- `mths_since_last_delinq`, `mths_since_last_record`, `mths_since_recent_bc_dlq`, `mths_since_last_major_derog`, and `mths_since_recent_revol_delinq` are missing not because of a data quality issue, but because the credit event never happened. For example, `mths_since_last_delinq` is empty for borrowers who have never been delinquent. Missing here is meaningful signal since a borrower with no delinquency history is fundamentally different from one who was delinquent 6 months ago. These columns will not be dropped. Instead, missing values will be filled with a large sentinel value (999) and a binary `_missing` indicator column will be added for each.


In [56]:
# Display moderate missingness columns
moderate_missingness = data_dictionary_summary[data_dictionary_summary['Missingness Label'] == 'Moderate (10-49%)']
print("Count of columns with moderate missingness (10-49%):", len(moderate_missingness))
print("Columns with moderate missingness (10-49%):")
moderate_missingness


Count of columns with moderate missingness (10-49%): 15
Columns with moderate missingness (10-49%):


,Variable,Description,Missing Values,Percentage (%),Missingness Label
28,il_util,Ratio of total current balance to high credit/credit limit on all install acct,1068883,47.281042,Moderate (10-49%)
53,mths_since_rcnt_il,Months since most recent installment accounts opened,909957,40.251099,Moderate (10-49%)
3,all_util,Balance to credit limit on all trades,866381,38.323555,Moderate (10-49%)
31,inq_last_12m,Number of credit inquiries in past 12 months,866163,38.313912,Moderate (10-49%)
103,total_cu_tl,Number of finance trades,866163,38.313912,Moderate (10-49%)
74,open_acc_6m,Number of open trades in last 6 months,866163,38.313912,Moderate (10-49%)
76,open_il_24m,Number of installment accounts opened in past 24 months,866162,38.313868,Moderate (10-49%)
78,open_rv_12m,Number of revolving trades opened in past 12 months,866162,38.313868,Moderate (10-49%)
79,open_rv_24m,Number of revolving trades opened in past 24 months,866162,38.313868,Moderate (10-49%)
43,max_bal_bc,Maximum current balance owed on all revolving accounts,866162,38.313868,Moderate (10-49%)


Note:
- `il_util`, `mths_since_rcnt_il`, `all_util`, `inq_last_12m`, `total_cu_tl`, `open_acc_6m`, `open_il_24m`, `open_il_12m`, `open_rv_12m`, `open_rv_24m`, `max_bal_bc`, `total_bal_il`, `open_act_il`, and `inq_fi` are credit bureau features that were introduced later in Lending Club's history. Older loans from 2007–2012 do not have them, so the missingness is temporal rather than random. These columns will be kept but flagged for the team to review before modeling.

- `mths_since_recent_inq` is missing because the borrower had no recent credit inquiry, not because of a data quality issue. Missing here is meaningful signal. These values will be filled with a median and a binary `_missing` indicator column will be added.

In [59]:
# Display low missingness columns
low_missingness = data_dictionary_summary[data_dictionary_summary['Missingness Label'] == 'Low (<10%)']
print("Count of columns with low missingness (<10%):", len(low_missingness))
print("Columns with low missingness (<10%):")
low_missingness

Count of columns with low missingness (<10%): 91
Columns with low missingness (<10%):


,Variable,Description,Missing Values,Percentage (%),Missingness Label
20,emp_title,The job title supplied by the Borrower when applying for the loan.*,167002,7.387178,Low (<10%)
69,num_tl_120dpd_2m,Number of accounts currently 120 days past due (updated in past 2 months),153690,6.798334,Low (<10%)
19,emp_length,Employment length in years. Possible values are between 0 and 10 where 0 means less than one year and 10 means ten or more years.,146940,6.499754,Low (<10%)
45,mo_sin_old_il_acct,Months since oldest bank installment account opened,139104,6.153136,Low (<10%)
9,bc_util,Ratio of total current balance to high credit/credit limit for all bankcard accounts.,76104,3.366389,Low (<10%)
...,...,...,...,...,...
94,term,The number of payments on the loan. Values are in months and can be either 36 or 60.,33,0.001460,Low (<10%)
92,sub_grade,LC assigned loan subgrade,33,0.001460,Low (<10%)
143,disbursement_method,"The method by which the borrower receives their loan. Possible values are: CASH, DIRECT_PAY",33,0.001460,Low (<10%)
128,hardship_flag,Flags whether or not the borrower is on a hardship plan,33,0.001460,Low (<10%)


Note:
- `emp_title` is the job title supplied by the borrower. It is free text with thousands of unique values and is not usable without NLP, which is out of scope for this project. It may be dropped, or be grouped based on job categories using NLP.

- `emp_length` is the employment length in years. Missing values here are ambiguous: it could mean unemployed, self-employed, or simply not provided. This will be imputed with the mode and a `_missing` indicator column will be added. The assumption will be documented before imputing.

- `num_tl_120dpd_2m` is the number of accounts currently 120 days past due, updated in the past 2 months. Note: If a borrower has 3 credit cards and 2 of them are 120+ days overdue, this value would be 2. Missingness here menas some borrowers had no accounts 120+ days past due (the event never happened).

- `mo_sin_old_il_acct` is the number of months since the oldest bank installment account was opened. Similar to the temporal missingness group, this may be missing for older loans that predate Lending Club's data collection of this field.

- `sub_grade` is Lending Club's own assigned loan subgrade, derived from the same applicant data the model will use. Keeping it would let the model learn Lending Club's risk assessment rather than build its own. It will be dropped.

- `hardship_flag` flags whether the borrower is on a hardship plan. This is post-origination information: it is updated after the loan is issued. It will be dropped.

- `debt_settlement_flag` flags whether a charged-off borrower is working with a debt settlement company. This is post-origination and a leakage risk — it only applies to borrowers who have already defaulted. It will be dropped.

- All remaining columns in this group have less than 3% missing values and will be imputed with the median, mode or something else.